src/prefect/utilities/asyncutils.py

```diff
@@ -32,14 +32,26 @@ def is_async_fn(
     func: Union[Callable[P, R], Callable[P, Awaitable[R]]]
 ) -> TypeGuard[Callable[P, Awaitable[R]]]:
     """
-    This wraps `iscoroutinefunction` with a `TypeGuard` such that we can perform a
-    conditional check and type checkers will narrow the expected type.
+    Returns `True` if a function returns a coroutine.
 
     See https://github.com/microsoft/pyright/issues/2142 for an example use
     """
+    while hasattr(func, "__wrapped__"):
+        func = func.__wrapped__
+
     return inspect.iscoroutinefunction(func)
 
 
+def is_async_gen_fn(func):
```

Comment:
```suggestion
def is_async_gen_fn(func: Callable) -> bool:
```

Reply:
I'm actually scared to try to reproduce the correct signature here which would be a type guard as above. I'm hoping by omitting it the type is inferred from the inner call which is a type guard.